eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,2478 @@
+	}
+	log.Debug("Persisted set of healing data", "bytes", common.StorageSize(batch.ValueSize()))
+}
+
+// processBytecodeHealResponse integrates an already validated bytecode response
+// into the healer tasks.
+func (s *Syncer) processBytecodeHealResponse(res *bytecodeHealResponse) {
+	for i, hash := range res.hashes {
+		node := res.codes[i]
+
+		// If the trie node was not delivered, reschedule it
+		if node == nil {
+			res.task.codeTasks[hash] = struct{}{}
+			continue
+		}
+		// Push the trie node into the state syncer
+		s.bytecodeHealSynced++
+		s.bytecodeHealBytes += common.StorageSize(len(node))
+
+		err := s.healer.scheduler.Process(trie.SyncResult{Hash: hash, Data: node})
+		switch err {
+		case nil:
+		case trie.ErrAlreadyProcessed:
+			s.bytecodeHealDups++
+		case trie.ErrNotRequested:
+			s.bytecodeHealNops++
+		default:
+			log.Error("Invalid bytecode processed", "hash", hash, "err", err)
+		}
+	}
+	batch := s.db.NewBatch()
+	if err := s.healer.scheduler.Commit(batch); err != nil {
+		log.Error("Failed to commit healing data", "err", err)
+	}
+	if err := batch.Write(); err != nil {
+		log.Crit("Failed to persist healing data", "err", err)
+	}
+	log.Debug("Persisted set of healing data", "bytes", common.StorageSize(batch.ValueSize()))
+}
+
+// forwardAccountTask takes a filled account task and persists anything available
+// into the database, after which it forwards the next account marker so that the
+// task's next chunk may be filled.
+func (s *Syncer) forwardAccountTask(task *accountTask) {
+	// Remove any pending delivery
+	res := task.res
+	if res == nil {
+		return // nothing to forward
+	}
+	task.res = nil
+
+	// Iterate over all the accounts and gather all the incomplete trie nodes. A
+	// node is incomplete if we haven't yet filled it (sync was interrupted), or
+	// if we filled it in multiple chunks (storage trie), in which case the few
+	// nodes on the chunk boundaries are missing.
+	incompletes := light.NewNodeSet()
+	for i := range res.accounts {
+		// If the filling was interrupted, mark everything after as incomplete
+		if task.needCode[i] || task.needState[i] {
+			for j := i; j < len(res.accounts); j++ {
+				if err := res.trie.Prove(res.hashes[j][:], 0, incompletes); err != nil {
+					panic(err) // Account range was already proven, what happened
+				}
+			}
+			break
+		}
+		// Filling not interrupted until this point, mark incomplete if needs healing
+		if task.needHeal[i] {
+			if err := res.trie.Prove(res.hashes[i][:], 0, incompletes); err != nil {
+				panic(err) // Account range was already proven, what happened
+			}
+		}
+	}
+	// Persist every finalized trie node that's not on the boundary
+	batch := s.db.NewBatch()
+
+	var (
+		nodes   int
+		skipped int
+		bytes   common.StorageSize
+	)
+	it := res.nodes.NewIterator(nil, nil)
+	for it.Next() {
+		// Boundary nodes are not written, since they are incomplete
+		if _, ok := res.bounds[common.BytesToHash(it.Key())]; ok {
+			skipped++
+			continue
+		}
```

Comment:
I think `bounds` are not properly vetted, a proof with (valid) leafs stuffed into the proof will prevent them from being persisted.

Reply:
It seems that during `VerifyProof`, we at one point do 
```
	// Remove all internal references. All the removed parts should
	// be re-filled(or re-constructed) by the given leaves range.
	if err := unsetInternal(root, firstKey, lastKey); err != nil {
		return nil, nil, false, err
	}
```
At this point, the `root` contains _exactly only_ the boundary nodes. A bit later, we fill it with leaves, and thus lose it again. And instead the outside code just uses the `proof` nodes as "boundary", because why not. 

If we were to copy the nodes after having done `unsetInternal`, and either 

- return it from `VerifyProof`, or, 
- verify the proof nodes against it, and bail if `proof` is bloated, 

Then we would get around this, I think. cc @rjl493456442

Reply:
On second thought, that wouldn't quite work, because the `unsetInternal`:

```
 it can happen that some fullnodes only have one child which is disallowed.
```

But it may still be possible to use the work performed there...